modules/core/ante/ante.go

```diff
@@ -29,25 +29,25 @@ func (ad AnteDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, ne
 		for _, m := range tx.GetMsgs() {
 			switch msg := m.(type) {
 			case *channeltypes.MsgRecvPacket:
-				if _, found := ad.k.GetPacketReceipt(ctx, msg.Packet.GetDestPort(), msg.Packet.GetDestChannel(), msg.Packet.GetSequence()); found {
+				if response, err := ad.k.RecvPacket(sdk.WrapSDKContext(ctx), msg); err == nil && response.Result == channeltypes.NOOP {
```

Comment:
I'm unsure if the behaviour here is correct. From my understanding (and this should be documented in the godoc above)

We are executing and committing state changes to the ctx CheckTx state. Thus if two redundant relays occur in the same block, the first state changes is committed and the second attempt would NOOP. This sound correct.

What isn't accounted for is that this AnteHandle check only rejects the tx if all relays are redundant, but the context is being committed for each single attempt (even if the other messages fail). So I could submit a tx with (recvMsg, timeoutMsg). My recvMsg is valid and my timeoutMsg is invalid. If my tx gets submitted first, any relayers using this ante handle would become blocked from submitting a relay for the recvMsg (unless they submit a tx without non-redundant messages) 

I think we need to cache this context. If an error occurs in any message the context is discarded and the next loop continues. If all messages succeed, the cache is written

Reply:
Wait, wouldn't it just work to say: If an error is returned by an message, then reject the tx?

Reply:
yup. Good idea! Not used to being able to run the entire tx in the ante handler